src/backend/distributed/utils/shard_utils.c

```diff
@@ -55,3 +56,23 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
 
 	return longestShardName;
 }
+
+
+/*
+ * GetLongestShardName is a utility function that creates a hypothetical shard
+ * name for a table that is not distributed yet.
+ *
+ * There is no easy way to create the names of shards before those shards are created
+ * because sequence manipulation in PostgreSQL is not transactional. Therefore we assume
+ * that the shardId values of a partition and those of its parent will be adjacent.
+ */
+char *
+GetLongestHypotheticalShardName(Oid parentTableOid, char *relationName)
```

Comment:
I don't want to over-engineer this, but parent table's shard could become outdated pretty quickly. So, assume we create 1 partition per day with shardCount 128, and hence the parent's largest shardId is could be too low after few days.

Ok, we don't want to call `GetNextShardId()`, but can we add `GetCurrentShardId()` and then get parents `` to generate the shardName:

```C
		CitusTableCacheEntry *parentCache = GetCitusTableCacheEntry(parentRelId);
        int shardCount = parentCache->shardIntervalArrayLength;
      
      int maxShardCount  = GetCurrentShardId()  + shardCount;
      char *newPartitionsName = pstrdup(get_rel_name(partitionId));
     return AppendShardIdToName(&newPartitionsName, maxShardCount);
```

Please first verify if this approach works fine, I just come up with this and didn't have much time to verify.

Reply:
AFAIK there is no easy way to get the last value generated by a sequence. `currval(regclass)` and `lastval()` gives errors if the `nextval()` function was not used earlier in the sessions.

> currval ( regclass ) → bigintReturns the value most recently obtained by nextval for this sequence in the current session. (An error is reported if nextval has never been called for this sequence in this session.) Because this is returning a session-local value, it gives a predictable answer whether or not other sessions have executed nextval since the current session did.This function requires USAGE or SELECT privilege on the sequence.
>
> lastval() → bigintReturns the value most recently returned by nextval in the current session. This function is identical to currval, except that instead of taking the sequence name as an argument it refers to whichever sequence nextval was most recently applied to in the current session. It is an error to call lastval if nextval has not yet been called in the current session.This function requires USAGE or SELECT privilege on the last used sequence.

I can however, iterate over all the shard values in the metadata, and return `1 + maxShardId` and IMO this would be overengineering.

See https://www.postgresql.org/docs/13/functions-sequence.html for the docs

Reply:
Can we use `pg_sequences`?
```SQL
select pg_sequence_last_value('pg_catalog.pg_dist_shardid_seq');
```

Because the current approach feels fragile and/or not very consistent. Parent's shardId is mostly likely not useful in many cases. What do you think?

Reply:
fwiw, `select last_value from pg_dist_shardid_seq` also gives the latest value, but we should probably just use the underlying C functions.

Reply:
`pg_sequence_last_value` returns NULL if the sequence never generated any numbers after the last reset.

I think it is safe to assert the sequence `is_called` flag will be true and it is safe to read its' last_value.

I can use the C function for `pg_sequence_last_value` to make this happen.

Reply:
I have 2 problems with using `pg_sequence_last_value` C function.
1. It is not declared in any header, and it uses several static functions in the same file. I don't think it is wise to copy the contents of those functions to our codebase as it is not easy to maintain.
2. It can potentially return `NULL` on several occasions:
  - the sequence is never used before
  - the sequence is never used after a command that set its `is_called` flag to false:
  	- `ALTER SEQUENCE .. RESTART`
  	- `SELECT setval( <regclass>, <bigint> , false )` that can set the `is_called` flag via the third parameter that is optional

```
hanefi=# SELECT * FROM pg_dist_shardid_seq ;
 last_value | log_cnt | is_called
------------+---------+-----------
     102008 |       0 | f
(1 row)

hanefi=# SELECT pg_sequence_last_value('pg_catalog.pg_dist_shardid_seq');
 pg_sequence_last_value
------------------------

(1 row)
```

My thinking is that it is better to run a SQL command (`SELECT last_value from pg_dist_shardid_seq`) in here and not use the underlying C functions to make sure the codebase is easier to maintain.